### PR TITLE
Get previous minor release for core components e2e tests

### PR DESCRIPTION
**Pull request by taneyland**
*Issue #, if available:*

*Description of changes:*
Methods to get the proper version of the previous minor release for the upcoming e2e tests for core components upgrades

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by abhay-krishna** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/aws/eks-anywhere/pull/618#" title="Author self-approved">taneyland</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=aws%2Feks-anywhere).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/aws/eks-anywhere/blob/main/OWNERS)~~ [taneyland]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by abhay-krishna** [comment]
@taneyland: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
eks-anywhere-e2e-presubmit | bb35cd6025fb99f02fb06bc835ec847c89370588 | [link](https://prow.eks.amazonaws.com/view/s3/prowpresubmitsdataclusterstack-prowbucket7c73355c-vfwwxd2eb4gp/pr-logs/pull/aws_eks-anywhere/618/eks-anywhere-e2e-presubmit/1458558670005604352) | `/test eks-anywhere-e2e-presubmit`

[Full PR test history](https://prow.eks.amazonaws.com/pr-history?org=aws&repo=eks-anywhere&pr=618). [Your PR dashboard](https://prow.eks.amazonaws.com/pr?query=is:pr+state:open+author:taneyland).

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by abhay-krishna** [comment]
@taneyland: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
eks-anywhere-e2e-presubmit | 94c38f90853781b7282a9848560c34016bcadc2b | [link](https://prow.eks.amazonaws.com/view/s3/prowpresubmitsdataclusterstack-prowbucket7c73355c-vfwwxd2eb4gp/pr-logs/pull/aws_eks-anywhere/618/eks-anywhere-e2e-presubmit/1458562840435625984) | `/test eks-anywhere-e2e-presubmit`

[Full PR test history](https://prow.eks.amazonaws.com/pr-history?org=aws&repo=eks-anywhere&pr=618). [Your PR dashboard](https://prow.eks.amazonaws.com/pr?query=is:pr+state:open+author:taneyland).

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

File: pkg/semver/semver.go
Comment (g-gaston) [inline pkg/semver/semver.go:78]: I don't think this is right
What about v is `v1.2.0` and v2 is `v2.1.0`
I would recommend writing tests for these methods

File: pkg/semver/semver.go
Comment (g-gaston) [inline pkg/semver/semver.go:87]: I'm not sure I understand the naming of this method
Why not just do `v.Minor < v2.Minor` wherever is needed?

File: test/e2e/components_upgrade_test.go
Comment (g-gaston) [inline test/e2e/components_upgrade_test.go:43]: what is this path referring to? Are you planning on adding some code to download the binaries to those paths?

File: test/e2e/components_upgrade_test.go
Comment (g-gaston) [inline test/e2e/components_upgrade_test.go:59]: typo 
```suggestion
			latestPrevMinorRelease = &release
```

File: test/e2e/components_upgrade_test.go
Comment (g-gaston) [inline test/e2e/components_upgrade_test.go:58]: I'm not sure how this works
* `latestPrevMinorRelease` is of type `*releasev1alpha1.EksARelease`, how can it be compared to `releaseVersion`? Also, it's nil at the start
* I'm not sure the `IsPrevMinorVersion`part will work for all the scenarios. What if they belong to different mayor versions? This might be enough for now, but a more robust way to do it would be `releaseVersion < releaseBranchVersion && releaseVersion.Minor != releaseBranchVersion.Minor && releaseVersion > latestPrevMinorRelease` (pseudo code)

File: test/e2e/components_upgrade_test.go
Comment (g-gaston) [inline test/e2e/components_upgrade_test.go:57]: Why ignore the error?

**Review by g-gaston — COMMENTED** [review]
Looks good!

File: test/e2e/components_upgrade_test.go
Comment (taneyland) [reply test/e2e/components_upgrade_test.go]: Yeah, I was planning to add that in a separate PR when we add the actual tests, but I think it'll be best to take this out for now until we solidify the framework for these e2e tests.

File: test/e2e/components_upgrade_test.go
Comment (danbudris) [inline test/e2e/components_upgrade_test.go:1]: I think these functions could be generally useful to the e2e test suite, consider if we should put them in their own file independent of the `components_upgrade_test.go` file, perhaps in their own file in the Framework?

**Review by danbudris — APPROVED** [review]

**Comment by abhay-krishna** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/aws/eks-anywhere/pull/618#pullrequestreview-806599624" title="Approved">danbudris</a>*, *<a href="https://github.com/aws/eks-anywhere/pull/618#" title="Author self-approved">taneyland</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=aws%2Feks-anywhere).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/aws/eks-anywhere/blob/main/OWNERS)~~ [danbudris,taneyland]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Review by danbudris — APPROVED** [review]

**Comment by abhay-krishna** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/aws/eks-anywhere/pull/618#pullrequestreview-806645892" title="Approved">danbudris</a>*, *<a href="https://github.com/aws/eks-anywhere/pull/618#" title="Author self-approved">taneyland</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=aws%2Feks-anywhere).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/aws/eks-anywhere/blob/main/OWNERS)~~ [danbudris,taneyland]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: pkg/semver/semver.go
Comment (g-gaston) [inline pkg/semver/semver.go:79]: I like this method
however, I would vote to re-implement it without nested if clauses
I think it would help with readability
I'm ok if you prefer to do it in a separate PR

File: test/e2e/components_upgrade_test.go
Comment (g-gaston) [inline test/e2e/components_upgrade_test.go]: do we need to parse this again?
can't we do `latestPrevMinorReleaseVersion = releaseVersion`

File: pkg/semver/semver.go
Comment (taneyland) [reply pkg/semver/semver.go:79]: I opted to write it this way to avoid repetitive checks. For example, we would have to check that the major versions are not equal twice to satisfy both the `GreaterThan` and the `LessThan` case if it's broken into separate if statements. If that's preferred for readability, I can update it in my next PR that builds off of this one.

**Pull request merged**
